diff --git a/RunCPM/console.h b/RunCPM/console.h
index 7729d89..83ce797 100644
--- a/RunCPM/console.h
+++ b/RunCPM/console.h
@@ -89,9 +89,17 @@ void _streamioReset(void) {
 uint8 _chready(void) // Checks if there's a character ready for input
 {
 #ifdef STREAMIO
-    if (streamInputActive)
-        return 0xff;
-    // TODO: Consider adding/keeping _abort_if_kbd_eof() here.
+    if (streamInputActive) {
+        int c = fgetc(streamInputFile);
+        if (c == EOF) {
+            streamInputActive = FALSE;
+        } else {
+            ungetc(c, streamInputFile);
+            return 0xff;
+        }
+    }
+    if (!streamInputActive && streamInputFile == stdin)
+        return (_kbhit() ? 0xff : 0x00);
     _abort_if_kbd_eof();
 #endif
     return (_kbhit() ? 0xff : 0x00);
@@ -102,7 +110,8 @@ uint8 _getconNB(void) // Gets a character, non-blocking, no echo
 #ifdef STREAMIO
     if (streamInputActive)
         return _getStreamInChar();
-    // TODO: Consider adding/keeping _abort_if_kbd_eof() here.
+    if (!streamInputActive && streamInputFile == stdin)
+        return (_kbhit() ? _getch() : 0x00);
     _abort_if_kbd_eof();
 #endif
     return (_kbhit() ? _getch() : 0x00);
@@ -113,7 +122,8 @@ uint8 _getcon(void) // Gets a character, blocking, no echo
 #ifdef STREAMIO
     if (streamInputActive)
         return _getStreamInChar();
-    // TODO: Consider adding/keeping _abort_if_kbd_eof() here.
+    if (!streamInputActive && streamInputFile == stdin)
+        return _getch();
     _abort_if_kbd_eof();
 #endif
     return _getch();
@@ -124,7 +134,8 @@ uint8 _getconE(void) // Gets a character, blocking, with echo
 #ifdef STREAMIO
     if (streamInputActive)
         return _getStreamInCharEcho();
-    // TODO: Consider adding/keeping _abort_if_kbd_eof() here.
+    if (!streamInputActive && streamInputFile == stdin)
+        return _getche();
     _abort_if_kbd_eof();
 #endif
     return _getche();
